app/main.py

```diff
@@ -1,34 +1,99 @@
-        # Find the corresponding deck in the list
-        pass
+    def _validate_ship(self, start: tuple[int, int], end: tuple[int, int]) -> None:
+        if not (start[0] == end[0] or start[1] == end[1]):
+            raise ValueError("Ship must be placed either horizontally or vertically.")
+        if not self.decks:
+            raise ValueError("Invalid ship size.")
+
+    def get_deck(self, row: int, column: int) -> Deck | None:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
+
+        return None
 
     def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+        deck = self.get_deck(row, column)
+        if deck and deck.is_alive:
+            deck.hit()
+            self.is_drowned = all(not d.is_alive for d in self.decks)
+            return "Sunk!" if self.is_drowned else "Hit!"
+        return "Miss!"
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: list[tuple[tuple[int, int], tuple[int, int]]]):
+        self.field = {}
+        self.ships = []
+
+        for start, end in ships:
+            ship = Ship(start, end)
+            self.ships.append(ship)
+            for deck in ship.decks:
+                self.field[(deck.row, deck.column)] = ship
+
+    def fire(self, location: tuple[int, int]) -> str:
+        if location in self.field:
+            ship = self.field[location]
+            result = ship.fire(*location)
+            return result
+        return "Miss!"
+
+    def _validate_field(self) -> None:
+        ship_counts = {1: 0, 2: 0, 3: 0, 4: 0}
+        for ship in self.ships:
+            ship_length = len(ship.decks)
+            if ship_length in ship_counts:
+                ship_counts[ship_length] += 1
+            else:
+                raise ValueError(f"Invalid ship length: {ship_length}")
+
+            for deck in ship.decks:
+                for dx in (-1, 0, 1):
+                    for dy in (-1, 0, 1):
+                        neighbor = (deck.row + dx, deck.column + dy)
+                        if neighbor in self.field and self.field[neighbor] != ship:
+                            raise ValueError("Ships must not be adjacent.")
+
+        if ship_counts != {1: 4, 2: 3, 3: 2, 4: 1}:
+            raise ValueError("Field must contain 1 four-deck, 2 three-deck, 3 two-deck, and 4 single-deck ships.")
+
```

Comment:
The `_validate_field` method is defined but not called. You should call this method in the `Battleship` constructor after all ships have been added to ensure the field is set up correctly according to the game rules.